File: python/z3b/natural.py
# Copyright (c) 2013 The SAYCBridge Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

from core import suit
from z3b import purposes
from z3b.constraints import *
from z3b.model import *
from z3b.preconditions import *
from z3b.rule_compiler import Rule, categories
from z3b.prefer import Highest, LowestLevel
from core.call import Call


def copy_dict(d, keys):
    return {key: d.get(key) for key in keys}


points_for_sound_suited_bid_at_level = [
    #  0   1   2   3   4   5   6   7
    None, 16, 19, 22, 25, 28, 33, 37,
]


# A notrump grand slam wants 37 HIGH-CARD points (the slam chapter, p156: "notrump slams require power -- generally
# 32+ HCP for a small slam and 37 HCP for a grand slam").  Partner's minimum here is total
# points, and whenever partner has shown a five-card suit it carries a length point that takes
# no trick in notrump, so the entry is 38: 37 high cards plus that point.  Opposite a balanced
# partner (a 1N opener's minimum has no length) it is a point strict, and those hands reach a
# grand through Gerber or a quantitative raise anyway.
points_for_sound_notrump_bid_at_level = [
    #  0   1   2   3   4   5   6   7
    None, 19, 22, 25, 28, 30, 33, 38,
]


class WeHaveShownMorePointsThanThem(Precondition):
    def fits(self, history, call):
        return history.us.min_points > history.them.min_points


class SufficientCombinedPoints(Constraint):
    """Partner's minimum (total points: with length or, after a raise, support points) plus this
    hand's high-card points reach the table's number -- the booklet's own arithmetic for the hand
    that has not revalued (a raise counts support points through MinimumCombinedSupportPoints)."""
    def tables(self):
        return points_for_sound_suited_bid_at_level, points_for_sound_notrump_bid_at_level

    def expr(self, history, call):
        strain = call.strain
        suited, notrump = self.tables()
        if strain == suit.NOTRUMP:
            min_points = notrump[call.level]
        else:
            assert strain in suit.SUITS, "%s not in %s" % (strain, suit.SUITS)
            min_points = suited[call.level]
        implied = max(0, min_points - history.partner.min_points)
        # Once both hands have agreed a suit the fit is known and shortness counts on both
        # sides: the bid is valued in support points for that suit, the way partner reads it.
        # A first raise stays on hcp plus length (the corpus: 3H, not 4H, on nine with a
        # doubleton after Stayman).
        if (strain in suit.SUITS and call.level <= 5
                and history.bid_suit_naturally(strain, positions.Partner)
                and history.bid_suit_naturally(strain, positions.Me)):
            return support_points_expr_for_suit(strain) >= implied
        return points >= implied


class SufficientCombinedLength(MinimumCombinedLength):
    def __init__(self):
        MinimumCombinedLength.__init__(self, 8)

    def expr(self, history, call):
        strain = call.strain
        if strain == suit.NOTRUMP:
            return NO_CONSTRAINTS
        return MinimumCombinedLength.expr(self, history, call)


class LengthSatisfiesLawOfTotalTricks(Constraint):
    def expr(self, history, call):
        # Written forward: level = partner_min + my_min - 6
        my_count = call.level + 6 - history.partner.min_length(call.strain)
        return expr_for_suit(call.strain) >= my_count


def _natural_suited_possible(call):
    if call.level >= 6:
        return ["Slam"]
    if call.level == 5 or (call.level == 4 and call.strain.char in "HS"):
        return ["Game", "SupportMajors"] if call.strain.char in "HS" else ["Game"]
    return ["RebidSuit", "Support", "AskLater", "Discovery"]


def natural_suited_purpose(history, call):
    """A natural suit bid is a slam or a game at those levels; below game it raises
    partner's suit, rebids our own, or discovers a new one."""
    if call.level >= 6:
        return "Slam"
    game = call.level == 5 or (call.level == 4 and call.strain.char in "HS")
    first = history.first_natural_bidder(call.strain)
    if first == positions.Me:
        return "Game" if game else "RebidSuit"  # our own suit, whether or not partner raised it
    if first == positions.Partner and call.strain.char in "HS" and call.level <= 4:
        return "SupportMajors"  # raising partner's major to game is still support: it beats exploring
    if game:
        return "Game"  # a minor game competes with 3N as a game
    if first == positions.Partner:
        return "SupportMinors"
    if any(history.bid_suit_naturally(s, positions.Me) and history.bid_suit_naturally(s, positions.Partner)
           for s in suit.SUITS):
        return "AskLater"  # a new suit once we have agreed one is a try, not a search for a fit
    return "Discovery"


def new_suit_purpose(history, call):
    """A new suit is discovery; a four-card minor shown at the two level or above waits
    behind a six-card rebid (a fifth card promotes it, see the rules' conditional purposes)."""
    if call.strain.char in "HS":
        return "MajorDiscovery"
    if call.level == 1:
        return "MinorDiscovery"
    return "MinorDiscoveryWithFour"


new_minor_with_five = [(MinLength(5), "MinorDiscovery", "MinorDiscoveryWithFour")]


def natural_notrump_purpose(history, call):
    if call.level >= 6:
        return "Slam"
    if call.level == 3:
        return "Game"
    return "CharacterizeStrength"


def law_of_total_tricks_purpose(history, call):
    """The law raises partner's suit to the level of the fit; anything else is competing."""
    if history.bid_suit_naturally(call.strain, positions.Partner):
        return "Support"
    return "Compete"


class Natural(Rule):
    category = categories.Natural


class SoundNaturalBid(Natural):
    shared_constraints = [
        SufficientCombinedLength(), SufficientCombinedPoints()]


# A natural suit bid: the slam the hand is worth (the higher first), else a game, else the
# lowest sufficient level, the higher suit within a level.
natural_suited_preference = [
    Highest(*Call.suited_names_between('6C', '7S')),
    LowestLevel('4H', '4S', '5C', '5D'),
    LowestLevel(*[name for name in Call.suited_names_between('2C', '5S') if name not in ('4H', '4S', '5C', '5D')]),
]


class NaturalSuited(SoundNaturalBid):
    """A natural suit bid: a raise of partner's suit, a rebid of our own, a new suit, a game
    or a slam.  The backstop of its purposes: any convention that applies says more."""
    purpose = natural_suited_purpose
    conditional_purposes = [(minor_raise_before_notrump, "SupportMinorWithFour", "SupportMinors"), (minor_raise_with_five, "SupportMinorWithFive", "SupportMinors")]  # see constraints.minor_raise_before_notrump
    fallback = 1
    preconditions = [
        InvertedPrecondition(LastBidHasAnnotation(
            positions.Partner, annotations.Preemptive)),
        WeHaveShownMorePointsThanThem(),
        PartnerHasAtLeastLengthInSuit(1),
    ]
    call_names = Call.suited_names_between('2C', '7S')
    prefer = natural_suited_preference


class LawOfTotalTricks(Rule):
    purpose = law_of_total_tricks_purpose
    conditional_purposes = [(MinLength(4), "SupportMinorWithFour", "SupportMinors")]
    preconditions = [
        # FIXME: This should only apply over weak bids (only when NaturalSuited does not)?
        PartnerHasAtLeastLengthInSuit(1),
        # A backup for competitive auctions, not a way past partner's signoff.
        InvertedPrecondition(LastBidHasAnnotation(positions.Partner, annotations.Signoff)),
    ]
    call_names = Call.suited_names_between('2C', '5D')
    shared_constraints = LengthSatisfiesLawOfTotalTricks()
    fallback = 2  # the backup raise: when neither a convention's raise nor a natural raise applies
    prefer = [Highest('4H', '4S', '5C', '5D'), Highest(*Call.suited_names_between('2C', '5D'))]  # a game, else the level of the fit
    category = categories.LawOfTotalTricks


class SufficientStoppers(Constraint):
    def _is_jump(self, last_contract, call):
        if not last_contract:
            return call.level > 1
        assert call.strain == suit.NOTRUMP
        if last_contract.strain == suit.NOTRUMP:
            return call.level > last_contract.level + 1
        return call.level > last_contract.level

    def expr(self, history, call):
        if self._is_jump(history.last_contract, call) and not history.partner.is_balanced:
            return StoppersInOpponentsSuits().expr(history, call)
        return NO_CONSTRAINTS


class NaturalNotrump(SoundNaturalBid):
    purpose = natural_notrump_purpose
    # A balanced hand's invitational 2N is its limit bid, before a raise of a minor or a suit
    # rebid; with shape those come first.
    conditional_purposes_per_call = {'2N': [(ConstraintAnd(balanced, NotEnoughForGame()), "BalancedLimit", "CharacterizeStrength")]}
    preconditions = WeHaveShownMorePointsThanThem()
    call_names = Call.notrump_names_between('1N', '7N')
    shared_constraints = SufficientStoppers()
    fallback = 1  # the backstop of its purposes: any convention's notrump call says more
    # The slam the hand is worth, else the lowest sufficient level (the Game preference
    # tells a stopped 3N from an unstopped one).
    prefer = [Highest('6N', '7N'), LowestLevel('1N', '2N', '3N', '4N', '5N')]


class DefaultPass(Rule):
    purpose = "Forced"
    preconditions = InvertedPrecondition(ForcedToBid())
    call_names = 'P'
    shared_constraints = NO_CONSTRAINTS
    category = categories.DefaultPass
    prefer = []
    fallback = 1  # the pass of last resort: any forced minimum call comes first



class NaturalPass(Rule):
    preconditions = [
        LastBidWas(positions.RHO, 'P'),
        # Natural passes do not apply when preempting.
        WeHaveShownMorePointsThanThem(),
        InvertedPrecondition(ForcedToBid()),
    ]
    call_names = 'P'
    category = categories.NaturalPass


class NaturalPassWithFit(NaturalPass):
    preconditions = [
        LastBidHasSuit(positions.Partner),
        InvertedPrecondition(LastBidHasAnnotation(
            positions.Partner, annotations.Artificial)),
        HaveFit(),
    ]
    shared_constraints = MinimumCombinedLength(7, use_partners_last_suit=True)


class SuitGameIsRemote(NaturalPassWithFit):
    purpose = "CharacterizeStrength"
    preconditions = LastBidWasBelowGame()
    # FIXME: Shouldn't this be support points?
    shared_constraints = MaximumCombinedPoints(24)
    prefer = []
    fallback = 2  # passing is what is left when no natural bid applies either


class SuitSlamIsRemote(NaturalPassWithFit):
    purpose = "Enough"
    preconditions = [
        LastBidWasGameOrAbove(),
        LastBidWasBelowSlam(),
        InvertedPrecondition(LastBidHasStrain(positions.Partner, suit.NOTRUMP))
    ]
    shared_constraints = MaximumCombinedPointsOppositeMinimum(32)
    prefer = []
    fallback = 2  # passing is what is left when no natural bid applies either


class NotrumpSlamIsRemote(NaturalPass):
    purpose = "Game"  # passing 3N is choosing a game: only a major game (the Game preference) is better
    preconditions = [
        LastBidHasStrain(positions.Partner, suit.NOTRUMP),
        LastBidWasGameOrAbove(),
        LastBidWasBelowSlam(),
        # Partner's 5N (pick a slam / grand slam invitation) is forcing: never pass it.
        InvertedPrecondition(LastBidWas(positions.Partner, '5N')),
    ]
    shared_constraints = MaximumCombinedPointsOppositeMinimum(32)
    prefer = []
    fallback = 2  # passing is what is left when no natural bid applies either

File: src/engine/z3b/natural.ts
// Copyright (c) 2013 The SAYCBridge Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
//
// Ported from python/z3b/natural.py: the natural suited and notrump bids, the
// passes (the game and the slam are remote, the law of total tricks) and the
// helper constraints and preconditions they need.  The declaration convention
// is the header comment of rule_compiler.ts.
//
// cspell:ignore notrump preempting

import { assert } from "../core/assert";
import { Call } from "../core/call";
import { NOTRUMP, SUITS } from "../core/suit";
import {
  Constraint,
  ConstraintAnd,
  MaximumCombinedPoints,
  MaximumCombinedPointsOppositeMinimum,
  MinimumCombinedLength,
  MinLength,
  minorRaiseBeforeNotrump,
  minorRaiseWithFive,
  NotEnoughForGame,
  StoppersInOpponentsSuits,
} from "./constraints";
import type { History } from "./history";
import {
  balanced,
  exprForSuit,
  NO_CONSTRAINTS,
  points,
  positions,
  supportPointsExprForSuit,
} from "./model";
import {
  annotations,
  ForcedToBid,
  HaveFit,
  InvertedPrecondition,
  LastBidHasAnnotation,
  LastBidHasStrain,
  LastBidHasSuit,
  LastBidWas,
  LastBidWasBelowGame,
  LastBidWasBelowSlam,
  LastBidWasGameOrAbove,
  PartnerHasAtLeastLengthInSuit,
  Precondition,
} from "./preconditions";
import { Highest, LowestLevel } from "./prefer";
import {
  categories,
  type ConditionalPurpose,
  Rule,
  rule,
  type RuleClass,
} from "./rule_compiler";
import type { Expr } from "./z3";

export function copyDict<T>(
  d: Readonly<Record<string, T>>,
  keys: readonly string[],
): Record<string, T | null> {
  return Object.fromEntries(keys.map((key) => [key, d[key] ?? null]));
}

export const pointsForSoundSuitedBidAtLevel: (number | null)[] = [
  //  0   1   2   3   4   5   6   7
  null,
  16,
  19,
  22,
  25,
  28,
  33,
  37,
];

// A notrump grand slam wants 37 HIGH-CARD points (the slam chapter, p156: "notrump slams require power -- generally
// 32+ HCP for a small slam and 37 HCP for a grand slam").  Partner's minimum here is total
// points, and whenever partner has shown a five-card suit it carries a length point that takes
// no trick in notrump, so the entry is 38: 37 high cards plus that point.  Opposite a balanced
// partner (a 1N opener's minimum has no length) it is a point strict, and those hands reach a
// grand through Gerber or a quantitative raise anyway.
export const pointsForSoundNotrumpBidAtLevel: (number | null)[] = [
  //  0   1   2   3   4   5   6   7
  null,
  19,
  22,
  25,
  28,
  30,
  33,
  38,
];

export class WeHaveShownMorePointsThanThem extends Precondition {
  fits(history: History, call: Call): boolean {
    void call;
    return history.us.minPoints > history.them.minPoints;
  }
}

/**
 * Partner's minimum (total points: with length or, after a raise, support points) plus this
 * hand's high-card points reach the table's number -- the booklet's own arithmetic for the hand
 * that has not revalued (a raise counts support points through MinimumCombinedSupportPoints).
 */
export class SufficientCombinedPoints extends Constraint {
  tables(): [(number | null)[], (number | null)[]] {
    return [pointsForSoundSuitedBidAtLevel, pointsForSoundNotrumpBidAtLevel];
  }

  expr(history: History, call: Call): Expr {
    const strain = call.strain!;
    const [suited, notrump] = this.tables();
    let minPoints: number | null;
    if (strain === NOTRUMP) {
      minPoints = notrump[call.level!];
    } else {
      assert(
        SUITS.includes(strain),
        `${strain} not in ${SUITS.map((s) => String(s)).join(", ")}`,
      );
      minPoints = suited[call.level!];
    }
    const implied = Math.max(0, minPoints! - history.partner.minPoints);
    // Once both hands have agreed a suit the fit is known and shortness counts on both
    // sides: the bid is valued in support points for that suit, the way partner reads it.
    // A first raise stays on hcp plus length (the corpus: 3H, not 4H, on nine with a
    // doubleton after Stayman).
    if (
      SUITS.includes(strain) &&
      call.level! <= 5 &&
      history.bidSuitNaturally(strain, positions.Partner) &&
      history.bidSuitNaturally(strain, positions.Me)
    ) {
      return supportPointsExprForSuit(strain).ge(implied);
    }
    return points.ge(implied);
  }
}

export class SufficientCombinedLength extends MinimumCombinedLength {
  constructor() {
    super(8);
  }

  override expr(history: History, call: Call): Expr {
    const strain = call.strain!;
    if (strain === NOTRUMP) {
      return NO_CONSTRAINTS;
    }
    return super.expr(history, call);
  }
}

export class LengthSatisfiesLawOfTotalTricks extends Constraint {
  expr(history: History, call: Call): Expr {
    // Written forward: level = partner_min + my_min - 6
    const myCount = call.level! + 6 - history.partner.minLength(call.strain!);
    return exprForSuit(call.strain!).ge(myCount);
  }
}

export function _naturalSuitedPossible(call: Call): string[] {
  if (call.level! >= 6) {
    return ["Slam"];
  }
  if (
    call.level === 5 ||
    (call.level === 4 && "HS".includes(call.strain!.char))
  ) {
    return "HS".includes(call.strain!.char)
      ? ["Game", "SupportMajors"]
      : ["Game"];
  }
  return ["RebidSuit", "Support", "AskLater", "Discovery"];
}

/**
 * A natural suit bid is a slam or a game at those levels; below game it raises
 * partner's suit, rebids our own, or discovers a new one.
 */
export function naturalSuitedPurpose(history: History, call: Call): string {
  if (call.level! >= 6) {
    return "Slam";
  }
  const game =
    call.level === 5 || (call.level === 4 && "HS".includes(call.strain!.char));
  const first = history.firstNaturalBidder(call.strain!);
  if (first === positions.Me) {
    return game ? "Game" : "RebidSuit"; // our own suit, whether or not partner raised it
  }
  if (
    first === positions.Partner &&
    "HS".includes(call.strain!.char) &&
    call.level! <= 4
  ) {
    return "SupportMajors"; // raising partner's major to game is still support: it beats exploring
  }
  if (game) {
    return "Game"; // a minor game competes with 3N as a game
  }
  if (first === positions.Partner) {
    return "SupportMinors";
  }
  if (
    SUITS.some(
      (s) =>
        history.bidSuitNaturally(s, positions.Me) &&
        history.bidSuitNaturally(s, positions.Partner),
    )
  ) {
    return "AskLater"; // a new suit once we have agreed one is a try, not a search for a fit
  }
  return "Discovery";
}

/**
 * A new suit is discovery; a four-card minor shown at the two level or above waits
 * behind a six-card rebid (a fifth card promotes it, see the rules' conditional purposes).
 */
export function newSuitPurpose(history: History, call: Call): string {
  void history;
  if ("HS".includes(call.strain!.char)) {
    return "MajorDiscovery";
  }
  if (call.level === 1) {
    return "MinorDiscovery";
  }
  return "MinorDiscoveryWithFour";
}

export const newMinorWithFive: readonly ConditionalPurpose[] = [
  [new MinLength(5), "MinorDiscovery", "MinorDiscoveryWithFour"],
];

export function naturalNotrumpPurpose(history: History, call: Call): string {
  void history;
  if (call.level! >= 6) {
    return "Slam";
  }
  if (call.level === 3) {
    return "Game";
  }
  return "CharacterizeStrength";
}

/** The law raises partner's suit to the level of the fit; anything else is competing. */
export function lawOfTotalTricksPurpose(history: History, call: Call): string {
  if (history.bidSuitNaturally(call.strain!, positions.Partner)) {
    return "Support";
  }
  return "Compete";
}

export class Natural extends Rule {
  static override dsl = rule({
    category: categories.Natural,
  });
}

export class SoundNaturalBid extends Natural {
  static override dsl = rule({
    sharedConstraints: [
      new SufficientCombinedLength(),
      new SufficientCombinedPoints(),
    ],
  });
}

// A natural suit bid: the slam the hand is worth (the higher first), else a game, else the
// lowest sufficient level, the higher suit within a level.
export const naturalSuitedPreference = [
  new Highest(...Call.suitedNamesBetween("6C", "7S")),
  new LowestLevel("4H", "4S", "5C", "5D"),
  new LowestLevel(
    ...Call.suitedNamesBetween("2C", "5S").filter(
      (name) => !["4H", "4S", "5C", "5D"].includes(name),
    ),
  ),
];

/**
 * A natural suit bid: a raise of partner's suit, a rebid of our own, a new suit, a game
 * or a slam.  The backstop of its purposes: any convention that applies says more.
 */
export class NaturalSuited extends SoundNaturalBid {
  static override dsl = rule({
    purpose: naturalSuitedPurpose,
    // see constraints.minorRaiseBeforeNotrump
    conditionalPurposes: [
      [minorRaiseBeforeNotrump, "SupportMinorWithFour", "SupportMinors"],
      [minorRaiseWithFive, "SupportMinorWithFive", "SupportMinors"],
    ],
    fallback: 1,
    preconditions: [
      new InvertedPrecondition(
        new LastBidHasAnnotation(positions.Partner, annotations.Preemptive),
      ),
      new WeHaveShownMorePointsThanThem(),
      new PartnerHasAtLeastLengthInSuit(1),
    ],
    callNames: Call.suitedNamesBetween("2C", "7S"),
    prefer: naturalSuitedPreference,
  });
}

export class LawOfTotalTricks extends Rule {
  static override dsl = rule({
    purpose: lawOfTotalTricksPurpose,
    conditionalPurposes: [
      [new MinLength(4), "SupportMinorWithFour", "SupportMinors"],
    ],
    preconditions: [
      // FIXME: This should only apply over weak bids (only when NaturalSuited does not)?
      new PartnerHasAtLeastLengthInSuit(1),
      // A backup for competitive auctions, not a way past partner's signoff.
      new InvertedPrecondition(
        new LastBidHasAnnotation(positions.Partner, annotations.Signoff),
      ),
    ],
    callNames: Call.suitedNamesBetween("2C", "5D"),
    sharedConstraints: new LengthSatisfiesLawOfTotalTricks(),
    fallback: 2, // the backup raise: when neither a convention's raise nor a natural raise applies
    // a game, else the level of the fit
    prefer: [
      new Highest("4H", "4S", "5C", "5D"),
      new Highest(...Call.suitedNamesBetween("2C", "5D")),
    ],
    category: categories.LawOfTotalTricks,
  });
}

export class SufficientStoppers extends Constraint {
  _isJump(lastContract: Call | null, call: Call): boolean {
    if (!lastContract) {
      return call.level! > 1;
    }
    assert(call.strain === NOTRUMP);
    if (lastContract.strain === NOTRUMP) {
      return call.level! > lastContract.level! + 1;
    }
    return call.level! > lastContract.level!;
  }

  expr(history: History, call: Call): Expr {
    if (
      this._isJump(history.lastContract, call) &&
      !history.partner.isBalanced
    ) {
      // Python passes the call too; the constraint ignores it.
      return new StoppersInOpponentsSuits().expr(history);
    }
    return NO_CONSTRAINTS;
  }
}

export class NaturalNotrump extends SoundNaturalBid {
  static override dsl = rule({
    purpose: naturalNotrumpPurpose,
    // A balanced hand's invitational 2N is its limit bid, before a raise of a minor or a suit
    // rebid; with shape those come first.
    conditionalPurposesPerCall: {
      "2N": [
        [
          new ConstraintAnd(balanced, new NotEnoughForGame()),
          "BalancedLimit",
          "CharacterizeStrength",
        ],
      ],
    },
    preconditions: new WeHaveShownMorePointsThanThem(),
    callNames: Call.notrumpNamesBetween("1N", "7N"),
    sharedConstraints: new SufficientStoppers(),
    fallback: 1, // the backstop of its purposes: any convention's notrump call says more
    // The slam the hand is worth, else the lowest sufficient level (the Game preference
    // tells a stopped 3N from an unstopped one).
    prefer: [
      new Highest("6N", "7N"),
      new LowestLevel("1N", "2N", "3N", "4N", "5N"),
    ],
  });
}

export class DefaultPass extends Rule {
  static override dsl = rule({
    purpose: "Forced",
    preconditions: new InvertedPrecondition(new ForcedToBid()),
    callNames: "P",
    sharedConstraints: NO_CONSTRAINTS,
    category: categories.DefaultPass,
    prefer: [],
    fallback: 1, // the pass of last resort: any forced minimum call comes first
  });
}

export class NaturalPass extends Rule {
  static override dsl = rule({
    preconditions: [
      new LastBidWas(positions.RHO, "P"),
      // Natural passes do not apply when preempting.
      new WeHaveShownMorePointsThanThem(),
      new InvertedPrecondition(new ForcedToBid()),
    ],
    callNames: "P",
    category: categories.NaturalPass,
  });
}

export class NaturalPassWithFit extends NaturalPass {
  static override dsl = rule({
    preconditions: [
      new LastBidHasSuit(positions.Partner),
      new InvertedPrecondition(
        new LastBidHasAnnotation(positions.Partner, annotations.Artificial),
      ),
      new HaveFit(),
    ],
    sharedConstraints: new MinimumCombinedLength(7, true),
  });
}

export class SuitGameIsRemote extends NaturalPassWithFit {
  static override dsl = rule({
    purpose: "CharacterizeStrength",
    preconditions: new LastBidWasBelowGame(),
    // FIXME: Shouldn't this be support points?
    sharedConstraints: new MaximumCombinedPoints(24),
    prefer: [],
    fallback: 2, // passing is what is left when no natural bid applies either
  });
}

export class SuitSlamIsRemote extends NaturalPassWithFit {
  static override dsl = rule({
    purpose: "Enough",
    preconditions: [
      new LastBidWasGameOrAbove(),
      new LastBidWasBelowSlam(),
      new InvertedPrecondition(
        new LastBidHasStrain(positions.Partner, NOTRUMP),
      ),
    ],
    sharedConstraints: new MaximumCombinedPointsOppositeMinimum(32),
    prefer: [],
    fallback: 2, // passing is what is left when no natural bid applies either
  });
}

export class NotrumpSlamIsRemote extends NaturalPass {
  static override dsl = rule({
    // passing 3N is choosing a game: only a major game (the Game preference) is better
    purpose: "Game",
    preconditions: [
      new LastBidHasStrain(positions.Partner, NOTRUMP),
      new LastBidWasGameOrAbove(),
      new LastBidWasBelowSlam(),
      // Partner's 5N (pick a slam / grand slam invitation) is forcing: never pass it.
      new InvertedPrecondition(new LastBidWas(positions.Partner, "5N")),
    ],
    sharedConstraints: new MaximumCombinedPointsOppositeMinimum(32),
    prefer: [],
    fallback: 2, // passing is what is left when no natural bid applies either
  });
}

/** The concrete rules of this module, by name (see sayc.ts). */
export const RULE_CLASSES: Readonly<Record<string, RuleClass>> = {
  NaturalSuited,
  LawOfTotalTricks,
  NaturalNotrump,
  DefaultPass,
  SuitGameIsRemote,
  SuitSlamIsRemote,
  NotrumpSlamIsRemote,
};
